lib/linkerd2-proxy-resolve/src/recover.rs

```diff
@@ -0,0 +1,351 @@
+            target: target.clone(),
+            recover: self.recover.clone(),
+            resolve: self.resolve.clone(),
+            state: State::Connecting {
+                future,
+                backoff: None,
+            },
+        };
+
+        Self::Future { inner: Some(inner) }
+    }
+}
+
+// === impl ResolveFuture ===
+
+impl<T, E, R> Future for ResolveFuture<T, E, R>
+where
+    T: fmt::Display + Clone,
+    R: resolve::Resolve<T>,
+    R::Endpoint: Clone + PartialEq,
+    E: Recover,
+{
+    type Item = Resolution<T, E, R>;
+    type Error = Error;
+
+    fn poll(&mut self) -> Poll<Self::Item, Self::Error> {
+        let inner = self.inner.as_mut().expect("polled after complete");
+        try_ready!(inner.poll_connected());
+
+        Ok(Async::Ready(Resolution {
+            inner: self.inner.take().expect("polled after complete"),
+            cache: Cache::default(),
+        }))
+    }
+}
+
+// === impl Resolution ===
+
+impl<T, E, R> resolve::Resolution for Resolution<T, E, R>
+where
+    T: fmt::Display + Clone,
+    R: resolve::Resolve<T>,
+    R::Endpoint: Clone + PartialEq,
+    E: Recover,
+{
+    type Endpoint = R::Endpoint;
+    type Error = Error;
+
+    fn poll(&mut self) -> Poll<Update<Self::Endpoint>, Self::Error> {
+        loop {
+            // If a previous reconnect left endpoints to be added, add them
+            // immediately.
+            if let Some(pending) = self.cache.take_pending() {
+                return Ok(pending.into());
+            }
+
+            if let State::Connected {
+                ref mut resolution,
+                ref mut initial,
+            } = self.inner.state
+            {
+                if let Some(update) = initial.take() {
+                    let update = self.cache.reconcile(update);
+                    return Ok(update.into());
+                }
+
+                match resolve::Resolution::poll(resolution) {
+                    Ok(Async::NotReady) => return Ok(Async::NotReady),
+                    Ok(Async::Ready(update)) => {
+                        let update = self.cache.process_update(update);
+                        return Ok(update.into());
+                    }
+                    Err(e) => {
+                        self.inner.state = State::Failed {
+                            error: Some(e.into()),
+                            backoff: None,
+                        };
+                    }
+                }
+            }
+
+            try_ready!(self.inner.poll_connected());
+        }
+    }
+}
+
+// === impl Inner ===
+
+impl<T, E, R> Inner<T, E, R>
+where
+    T: fmt::Display + Clone,
+    R: resolve::Resolve<T>,
+    R::Endpoint: Clone + PartialEq,
+    E: Recover,
+{
+    fn poll_connected(&mut self) -> Poll<(), Error> {
+        loop {
+            self.state = match self.state {
+                State::Disconnected { ref mut backoff } => {
+                    tracing::trace!(message = "connecting", target = %self.target);
```

Comment:
Nit: IMO, it would be better for the `Resolution` type to hold a tracing `Span` with the target that we `enter` every time `Resolution::poll` is called. That way, we wouldn't have to manually include the target in these events, and any events emitted by methods we call would _also_ be in that span.

Also, eventually, it would allow us to do time profiling for resolutions...

Reply:
If we're going to do this, that should be its own layer in a followup, but it doesn't seem high-priority at the moment, since this all gets put on a task with proper context

Reply:
Simplified all of this by making target-tracing an explicit stack element.

Reply:
aaand undid that, but removed the target from here, since it's expected to be already part of the instrumented context